lib/msal-node/src/client/ClientApplication.ts

```diff
@@ -279,33 +281,126 @@ export abstract class ClientApplication {
+            return await this.acquireCachedTokenSilent(
+                validRequest,
+                silentFlowClient,
+                clientConfiguration
+            );
+        } catch (error) {
+            if (this.tokenCache.persistence) {
+                try {
+                    await this.tokenCache.overwriteCache();
+                    this.logger.info("Searching again for a valid token");
+                    return await this.acquireCachedTokenSilent(
+                        validRequest,
+                        silentFlowClient,
+                        clientConfiguration
+                    );
+                } catch (overwriteError) {
+                    return this.handleRefreshTokenFlow(
+                        validRequest,
+                        clientConfiguration,
+                        overwriteError as Error
+                    );
+                }
+            }
+            return this.handleRefreshTokenFlow(
+                validRequest,
+                clientConfiguration,
+                error as Error
+            );
         }
     }
 
+    private async handleRefreshTokenFlow(
+        validRequest: CommonSilentFlowRequest,
+        clientConfiguration: ClientConfiguration,
+        error: Error
+    ): Promise<AuthenticationResult> {
+        if (
+            error instanceof ClientAuthError &&
+            error.errorCode === ClientAuthErrorCodes.tokenRefreshRequired
+        ) {
+            const refreshTokenClient = new RefreshTokenClient(
+                clientConfiguration
+            );
+            return refreshTokenClient.acquireTokenByRefreshToken(validRequest);
+        }
+        throw error;
+    }
+
+    private async acquireCachedTokenSilent(
+        validRequest: CommonSilentFlowRequest,
+        silentFlowClient: SilentFlowClient,
+        clientConfiguration: ClientConfiguration
+    ): Promise<AuthenticationResult> {
+        const [authResponse, cacheOutcome] =
+            await silentFlowClient.acquireCachedToken({
+                ...validRequest,
+                scopes: validRequest.scopes?.length
+                    ? validRequest.scopes
+                    : [...OIDC_DEFAULT_SCOPES],
+            });
+
+        if (cacheOutcome === CacheOutcome.PROACTIVELY_REFRESHED) {
+            this.logger.info(
+                "ClientApplication:acquireCachedTokenSilent - Cached access token's refreshOn property has been exceeded'. It's not expired, but must be refreshed."
+            );
+            // refresh the access token in the background
+            const refreshTokenClient = new RefreshTokenClient(
+                clientConfiguration
+            );
+
+            try {
+                await refreshTokenClient.acquireTokenByRefreshToken(
+                    validRequest
+                );
+            } catch {
+                // do nothing, this is running in the background and no action is to be taken upon success or failure
+            }
```

Comment:
I see you implemented my previous feedback here. 

After looking at it a second time, I think you can use try/finally so that there's not an empty catch?

```javascript
try {
  console.log('This is the try block.');
  // You can throw an error here if needed, like:
  // throw new Error('Something went wrong!');
} finally {
  console.log('This is the finally block. It always runs!');
}
```

Reply:
For some reason, Typescript doesn't recognize that the finally block will always run, so if I move the return statement in the finally block, I get an error that I have to change the return type of the function to `Promise<AuthenticationResult> | undefined`, which would require more changes in the code. I'm guessing this is why the original code also had the try/catch.

Reply:
I did some research and it appears putting a return in a finally block is bad practice anyways. Sorry!